bookkeeper-common/src/main/java/org/apache/bookkeeper/common/collections/RecyclableHashSet.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.bookkeeper.common.collections;
+
+import io.netty.util.Recycler.Handle;
+import java.util.HashSet;
+import org.apache.bookkeeper.common.util.Recyclable;
+
+/**
+ * A hash set which is recyclable.
+ */
+public final class RecyclableHashSet<T> extends HashSet<T> implements Recyclable {
```

Comment:
Making a HashSet recyclable wins you very little. A HashSet is really a HashMap<T, Boolean> underneath, so any lookups and puts will be generating non-recyled Cell objects. 

If we're going down this path, it would be better to use a ArrayList rather than a HashSet for Watcher, as it stores the objects directly. Insertion is O(n) if you want to ensure uniqueness, but the number of watchers is likely to be small enough that this would outperform a HashSet anyhow due to cache locality.

Reply:
it still saves you one allocation per lac update and long poll request. Based on current implementation, we can potentially use RecyclableArrayList, I will look into this.